src/plugins/unified_histogram/public/container/container.tsx

```diff
@@ -0,0 +1,150 @@
+ */
+export type UnifiedHistogramContainerProps = Pick<
+  UnifiedHistogramLayoutProps,
+  'className' | 'resizeRef' | 'appendHitsCounter' | 'children'
+>;
+
+/**
+ * The options used to initialize the container
+ */
+export type UnifiedHistogramInitializeOptions = UnifiedHistogramStateOptions &
+  Omit<LayoutProps, 'services'>;
+
+/**
+ * The uninitialized API exposed by the container
+ */
+export interface UnifiedHistogramUninitializedApi {
+  /**
+   * Whether the container has been initialized
+   */
+  initialized: false;
+  /**
+   * Initialize the container
+   */
+  initialize: (options: UnifiedHistogramInitializeOptions) => void;
+}
+
+/**
+ * The initialized API exposed by the container
+ */
+export interface UnifiedHistogramInitializedApi {
+  /**
+   * Whether the container has been initialized
+   */
+  initialized: true;
+  /**
+   * Get an observable to watch the container state, optionally providing a
+   * selector function which only gets triggered when the selected state changes
+   */
+  getState$: UnifiedHistogramStateService['getState$'];
+  /**
+   * Update the container state my providing a partial state object
+   */
+  updateState: UnifiedHistogramStateService['updateState'];
+  /**
+   * Manually trigger a refetch of the data
+   */
+  refetch: () => void;
+}
+
+/**
+ * The API exposed by the container
+ */
+export type UnifiedHistogramApi = UnifiedHistogramUninitializedApi | UnifiedHistogramInitializedApi;
+
+export const UnifiedHistogramContainer = forwardRef<
+  UnifiedHistogramApi,
+  UnifiedHistogramContainerProps
+>((containerProps, ref) => {
+  const [initialized, setInitialized] = useState(false);
+  const [layoutProps, setLayoutProps] = useState<LayoutProps>();
+  const [state, setState] = useState<UnifiedHistogramState>();
+  const [stateService, setStateService] = useState<UnifiedHistogramStateService>();
+  const stateProps = useStateProps({ state, stateService });
+  const input$ = useMemo(() => new Subject<UnifiedHistogramInputMessage>(), []);
+
+  useImperativeHandle(
+    ref,
+    () => ({
+      initialized,
+      initialize: (options: UnifiedHistogramInitializeOptions) => {
+        const { services, disableAutoFetching, disableTriggers, disabledActions } = options;
+
+        setLayoutProps({ services, disableAutoFetching, disableTriggers, disabledActions });
+        setStateService(new UnifiedHistogramStateService(options));
+        setInitialized(true);
+      },
+      getState$: (selector) => {
+        return stateService!.getState$(selector);
+      },
+      updateState: (stateUpdate) => {
+        stateService!.updateState(stateUpdate);
+      },
+      refetch: () => {
+        input$.next({ type: 'refetch' });
+      },
+    }),
+    [initialized, input$, stateService]
+  );
+
+  // Subscribe to state changes
+  useEffect(() => {
+    if (!stateService) {
+      return;
+    }
+
+    const subscription = stateService.getState$().subscribe(setState);
```

Comment:
I'd remove this and the related state useState and subscribe to state changes where they're needed AND only the subset of state thats needs.

Reply:
Removed this and switched to `useStateSelector` within the container.